### ISSUE TITLE
Async version of IMap.values() and IMap.values(Predicate)

### ISSUE THREAD

**Issue by elR1co** [hazelcast/hazelcast#17370]
Today `IMap.values()` and `IMap.values(Predicate)` calls are blocking.

I would like to use `IMap.values(Predicate)` in a Jet Pipeline, which is possible, but I need to declare it as nonCooperative, and will have an impact on the pipeline scalability.

Would it be possible to have an async (non-blocking) version for these calls ?

Thank you very much for all the hard work done !

**Comment by mmedenjak** [comment]
@elR1co thank you for the suggestion. I believe the functionality should be rather simple to add but since we're at the end of the release cycle and since we're heavily investing into our new SQL engine, we'll postpone work on this. Still, if you want to take a shot at it, please do, we can try and guide you.

**Comment by harshvvi** [comment]
@mmedenjak I would like to work on this and as It will be my first PR can you give a little insight how it should be done.

**Comment by mmedenjak** [comment]
Hi @harshvvi ! Although this issue will be simple in terms of not having to deal with many subsystems, it will be a bit involved since you'll have to decouple code which needs to run after the results have been fetched and before they have been returned to the user. Now, if we look at the code, both `IMap#values()` and `IMap#values(com.hazelcast.query.Predicate<K,V>)` end up calling `QueryEngineImpl#execute`. Here, there are three options. The first two will invoke remote operations and then call `QueryEngineImpl#addResultsOfPredicate`. This method is where we block the user thread. Instead, we should rearrange the code in so that after the remote invocation, we retrieve the `Future` and apply the rest of the transformations in a callback passed to something like `CompletionStage#thenApply`. The `CompletionStage` returned from the `thenApply`  is what we return to the user.
Regarding the third option (`runOnGivenPartition(adjustedQuery, target)`), there is a call to `.get();` inside the method which we can eliminate and return the future instead. The same applies here - if there was any processing to the result after the results have been retrieved, we'll need to do it in a callback passed to something like `thenApply`.

Basically, you need to collect the futures returned from remote invocations and apply the rest of the transformations in an asynchronous callback. The future returned to the user will have to be the one which has all of the transformations already applied.

Hope that made sense. A very very very simple example of this would be [here](https://github.com/hazelcast/hazelcast/blob/f6c9913b0f4c99997d8ac1b229f75e769121605c/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java#L721-L724). As you can see, there's a future returned from `registerListenerAsync`, then we apply some callback and return the resulting future to the user. In the case of `IMap.values`, the callback will be large and probably split into multiple methods.

**Comment by harshvvi** [comment]
Hi @mmedenjak, sorry for the late reply I'm new to java concurrency hence I was doing a little bit of research. So what I can understand is that (I might be completely wrong) instead of calling [addResultsOfPredicate](https://github.com/hazelcast/hazelcast/blob/c064a59b87a27378e1e36e5168db6234e2b2a6e7/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java#L197) from methods [doRunOnQueryThreads](https://github.com/hazelcast/hazelcast/blob/c064a59b87a27378e1e36e5168db6234e2b2a6e7/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java#L153) and [doRunOnPartitionThreads](https://github.com/hazelcast/hazelcast/blob/c064a59b87a27378e1e36e5168db6234e2b2a6e7/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java#L185) we can apply the transformation using a callback if valuesAsync is called and than return the completionStage.

**Comment by mmedenjak** [comment]
Yes, something like that. I think it'll get a bit complicated as different executions of `IMap.values()` end up doing different things after retrieving the initial results and you'll need to refactor the code a bit not to have a single monolithic method with all cases and different callbacks.

Also, I think you can have `IMap.values()` internally call `IMap.valuesAsync()` and just block until the future (completion stage) is finished. That way, both methods will use the same code and we don't have to duplicate it for sync and async variants.

**Comment by viliam-durina** [comment]
@Sm0keySa1m0n You can try to use SQL for this. It's not an async API, but at least it's streaming.
```sql
CREATE MAPPING my_map ... ;
SELECT this FROM my_map;
```
